sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala

```diff
@@ -785,7 +785,7 @@ class CachedTableSuite extends QueryTest with SQLTestUtils with SharedSparkSessi
       assert(getNumInMemoryRelations(ds) == 2)
 
       val cachedDs = sql(sqlText).cache()
-      assert(getNumInMemoryTablesRecursively(cachedDs.queryExecution.sparkPlan) == 3)
+      assert(getNumInMemoryTablesRecursively(cachedDs.queryExecution.sparkPlan) == 2)
```

Comment:
why this becomes 2?

Reply:
> why this becomes 2?

Since we rewrite it in Optimizer,  don't hit cache.

Reply:
previously we also rewrite it (to join) in the optimizer, right?

Reply:
> previously we also rewrite it (to join) in the optimizer, right?
**Before **
```
+- Filter NOT exists#230 []
   :  +- Project [c1#222]
   :     +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
   :           +- *(1) Project [value#219 AS c1#222]
   :              +- *(1) LocalTableScan [value#219]
   +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
         +- *(1) Project [value#219 AS c1#222]
            +- *(1) LocalTableScan [value#219]

InMemoryTableScan [c1#222]
   +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
         +- BroadcastNestedLoopJoin BuildRight, LeftAnti
            :- *(1) ColumnarToRow
            :  +- Scan In-memory table t1 [c1#222]
            :        +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
            :              +- *(1) Project [value#219 AS c1#222]
            :                 +- *(1) LocalTableScan [value#219]
            +- BroadcastExchange IdentityBroadcastMode, [id=#103]
               +- *(2) ColumnarToRow
                  +- Scan In-memory table t1
                        +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
                              +- *(1) Project [value#219 AS c1#222]
                                 +- *(1) LocalTableScan [value#219]
```

**Now**
```
Project [c1#222]
+- Filter NOT exists#230 []
   :  +- Project [c1#222]
   :     +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
   :           +- *(1) Project [value#219 AS c1#222]
   :              +- *(1) LocalTableScan [value#219]
   +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
         +- *(1) Project [value#219 AS c1#222]
            +- *(1) LocalTableScan [value#219]

InMemoryTableScan [c1#222]
   +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
         +- *(1) Filter isnull(Subquery scalar-subquery#253, [id=#101])
            :  +- Subquery scalar-subquery#253, [id=#101]
            :     +- CollectLimit 1
            :        +- *(1) Project [1 AS col#270]
            :           +- *(1) ColumnarToRow
            :              +- Scan In-memory table t1
            :                    +- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
            :                          +- *(1) Project [value#219 AS c1#222]
            :                             +- *(1) LocalTableScan [value#219]
            +- *(1) ColumnarToRow
               +- Scan In-memory table t1 [c1#222], [isnull(ReusedSubquery Subquery scalar-subquery#253, [id=#101])]
                     :- InMemoryRelation [c1#222], StorageLevel(disk, memory, deserialized, 1 replicas)
                     :     +- *(1) Project [value#219 AS c1#222]
                     :        +- *(1) LocalTableScan [value#219]
                     +- ReusedSubquery Subquery scalar-subquery#253, [id=#101]
```

Reply:
@cloud-fan
Just see current `SparkPlan`, it's because reuse subquery.

Reply:
ah i see. Can we update this test to create 2 tables t1 and t2? so that we won't hit subquery reuse.

Reply:
> ah i see. Can we update this test to create 2 tables t1 and t2? so that we won't hit subquery reuse.

Seem I make a mistake, `getNumInMemoryTablesRecursively` method can't find `InMemoryTableScanExec` in ScalaSubquery

Reply:
And seems  here shouldn't have this `ReuseSubquery`

Reply:
> And seems here shouldn't have this `ReuseSubquery`

Got it, filter push down to InMemoryTableScanExec. 
The problem is  `getNumInMemoryTablesRecursively` method can't find `InMemoryTableScanExec` in `ScalaSubquery`

Reply:
Just change `getNumInMemoryTablesRecursively()` method here? it's only been used in this place

Reply:
yea let's change `getNumInMemoryTablesRecursively`

Reply:
> yea let's change `getNumInMemoryTablesRecursively`

Changed.